File: src/interface.zig
const std = @import("std");
const rl = @import("raylib");
const Ssd1351 = @import("oled").Ssd1351;
const config = @import("config");

pub const WIDTH = 128;
pub const HEIGHT = 128;

var target: rl.RenderTexture2D = undefined;
var oled: Ssd1351 = undefined;

pub fn init() !void {
    // rl.setConfigFlags(.{ .window_resizable = true }); // commented because it looks weird with aerospace window manager
    rl.initWindow(512, 512, "LeDaw");
    oled = Ssd1351.init();

    target = try rl.loadRenderTexture(WIDTH, HEIGHT);
    rl.setTargetFPS(60);
    rl.setExitKey(.null); // ESC doesn't close program
}

pub fn deinit() void {
    rl.unloadRenderTexture(target);
    rl.closeWindow();

    oled.deinit();
}

pub fn preRender() void {
    rl.beginTextureMode(target);
    rl.clearBackground(rl.Color.black);
}

pub fn postRender() void {
    const screen_width = rl.getScreenWidth();
    const screen_height = rl.getScreenHeight();
    const square_len = @min(screen_width, screen_height);
    const pos_x: f32 = @floatFromInt(@divTrunc(screen_width - square_len, 2));
    const pos_y: f32 = @floatFromInt(@divTrunc(screen_height - square_len, 2));
    const square_len_f: f32 = @floatFromInt(square_len);

    rl.endTextureMode();
    rl.beginDrawing();
    defer rl.endDrawing();

    if (config.hw) {
        // copy target texture to oled
        const image = rl.loadImageFromTexture(target.texture);
        defer rl.unloadImage(image);
        const rgb8888: [*]const u8 = @ptrCast(image.data);
        var fb = rgb8888_to_rgb565(rgb8888);
        oled.show(&fb);
    } else {
        // render the 128x128 square
        rl.drawTexturePro(
            target.texture,
            rl.Rectangle{
                .x = 0,
                .y = 0,
                .width = @floatFromInt(target.texture.width),
                .height = -@as(f32, @floatFromInt(target.texture.height)), // flip Y
            },
            rl.Rectangle{
                .x = pos_x,
                .y = pos_y,
                .width = square_len_f,
                .height = square_len_f,
            },
            rl.Vector2{ .x = 0, .y = 0 },
            0.0,
            rl.Color.white,
        );
    }
}

// ---------------------------------------
// utility stuff
// ---------------------------------------

fn rgb8888_to_rgb565(px: []const u8) []u16 {
    var ans: [WIDTH * HEIGHT]u16 = undefined;
    var y: usize = 0;
    while (y < HEIGHT) : (y += 1) {
        const src_y = HEIGHT - 1 - y; // flip: GL bottom-origin -> OLED top-origin
        var x: usize = 0;
        while (x < WIDTH) : (x += 1) {
            const i = (src_y * WIDTH + x) * 4;
            const r = px[i];
            const g = px[i + 1];
            const b = px[i + 2];
            ans[y * WIDTH + x] = (@as(u16, r >> 3) << 11) | (@as(u16, g >> 2) << 5) | (b >> 3);
        }
    }
    return ans;
}

pub fn drawTextCentered(text: [:0]const u8, center_x: i32, center_y: i32, font_size: i32, color: rl.Color) void {
    const text_width = rl.measureText(text, font_size);
    rl.drawText(text, center_x - @divTrunc(text_width, 2), center_y - @divTrunc(font_size, 2), font_size, color);
}

pub fn toString(comptime T: type, value: T) [20:0]u8 {
    var buf: [20:0]u8 = .{0} ** 20;
    _ = std.fmt.bufPrint(&buf, "{d}", .{value}) catch unreachable;
    return buf;
}

pub fn shouldClose() bool {
    return rl.windowShouldClose();
}

pub const EventType = enum {
    key_press,
    key_release,
};

pub const Event = struct {
    type: EventType,
    key: rl.KeyboardKey,
};

pub const poll_keys = [_]rl.KeyboardKey{
    .a,             .b,         .c,         .d,          .e,          .f,           .g,            .h,             .i,        .j,         .k,          .l,           .m,
    .n,             .o,         .p,         .q,          .r,          .s,           .t,            .u,             .v,        .w,         .x,          .y,           .z,
    .zero,          .one,       .two,       .three,      .four,       .five,        .six,          .seven,         .eight,    .nine,      .escape,     .grave,       .minus,
    .equal,         .backspace, .tab,       .caps_lock,  .left_shift, .right_shift, .left_control, .right_control, .left_alt, .right_alt, .left_super, .right_super, .left_bracket,
    .right_bracket, .backslash, .semicolon, .apostrophe, .enter,      .comma,       .period,       .slash,         .space,    .up,        .down,       .left,        .right,
    .delete,
};

var poll_index: usize = 0;
var poll_phase: enum { press, release } = .press;

pub fn nextEvent() ?Event {
    while (true) {
        if (poll_phase == .press) {
            while (poll_index < poll_keys.len) {
                const key = poll_keys[poll_index];
                poll_index += 1;
                if (rl.isKeyPressed(key)) {
                    return .{ .type = .key_press, .key = key };
                }
            }
            poll_index = 0;
            poll_phase = .release;
        }
        if (poll_phase == .release) {
            while (poll_index < poll_keys.len) {
                const key = poll_keys[poll_index];
                poll_index += 1;
                if (rl.isKeyReleased(key)) {
                    return .{ .type = .key_release, .key = key };
                }
            }
            poll_index = 0;
            poll_phase = .press;
            return null;
        }
    }
}

File: app/src/interface.rs
//! Desktop interface helpers (raylib). The hardware/OLED path from `interface.zig` is
//! dropped; only the desktop render path remains (render to a 128x128 texture, then blit
//! scaled to the window — done in `main.rs`).
//!
//! Feature-gated behind `ui`; needs the `raylib` crate + system GL/X11 libs. Not built or
//! verified in the core workspace.

use raylib::prelude::*;

use crate::input::{Event, EventType, Key, POLL_KEYS};

pub const WIDTH: i32 = 128;
pub const HEIGHT: i32 = 128;

/// Map our raylib-free `Key` onto raylib's `KeyboardKey` for polling.
pub fn rl_key(k: Key) -> KeyboardKey {
    use Key::*;
    use raylib::consts::KeyboardKey::*;
    match k {
        A => KEY_A,
        B => KEY_B,
        C => KEY_C,
        D => KEY_D,
        E => KEY_E,
        F => KEY_F,
        G => KEY_G,
        H => KEY_H,
        I => KEY_I,
        J => KEY_J,
        K => KEY_K,
        L => KEY_L,
        M => KEY_M,
        N => KEY_N,
        O => KEY_O,
        P => KEY_P,
        Q => KEY_Q,
        R => KEY_R,
        S => KEY_S,
        T => KEY_T,
        U => KEY_U,
        V => KEY_V,
        W => KEY_W,
        X => KEY_X,
        Y => KEY_Y,
        Z => KEY_Z,
        Zero => KEY_ZERO,
        One => KEY_ONE,
        Two => KEY_TWO,
        Three => KEY_THREE,
        Four => KEY_FOUR,
        Five => KEY_FIVE,
        Six => KEY_SIX,
        Seven => KEY_SEVEN,
        Eight => KEY_EIGHT,
        Nine => KEY_NINE,
        Escape => KEY_ESCAPE,
        Grave => KEY_GRAVE,
        Minus => KEY_MINUS,
        Equal => KEY_EQUAL,
        Backspace => KEY_BACKSPACE,
        Tab => KEY_TAB,
        CapsLock => KEY_CAPS_LOCK,
        LeftShift => KEY_LEFT_SHIFT,
        RightShift => KEY_RIGHT_SHIFT,
        LeftControl => KEY_LEFT_CONTROL,
        RightControl => KEY_RIGHT_CONTROL,
        LeftAlt => KEY_LEFT_ALT,
        RightAlt => KEY_RIGHT_ALT,
        LeftSuper => KEY_LEFT_SUPER,
        RightSuper => KEY_RIGHT_SUPER,
        LeftBracket => KEY_LEFT_BRACKET,
        RightBracket => KEY_RIGHT_BRACKET,
        Backslash => KEY_BACKSLASH,
        Semicolon => KEY_SEMICOLON,
        Apostrophe => KEY_APOSTROPHE,
        Enter => KEY_ENTER,
        Comma => KEY_COMMA,
        Period => KEY_PERIOD,
        Slash => KEY_SLASH,
        Space => KEY_SPACE,
        Up => KEY_UP,
        Down => KEY_DOWN,
        Left => KEY_LEFT,
        Right => KEY_RIGHT,
        Delete => KEY_DELETE,
    }
}

/// Collect this frame's key presses then releases (mirrors `interface.nextEvent`'s order).
pub fn poll_events(rl: &RaylibHandle) -> Vec<Event> {
    use raylib::consts::KeyboardKey::*;
    let mut out = Vec::new();
    for &k in POLL_KEYS.iter() {
        if rl.is_key_pressed(rl_key(k)) {
            out.push(Event {
                ty: EventType::KeyPress,
                key: k,
                shift: rl.is_key_down(KEY_LEFT_SHIFT),
                meta: rl.is_key_down(KEY_LEFT_SUPER),
            });
        }
    }
    for &k in POLL_KEYS.iter() {
        if rl.is_key_released(rl_key(k)) {
            out.push(Event {
                ty: EventType::KeyRelease,
                key: k,
                shift: rl.is_key_down(KEY_LEFT_SHIFT),
                meta: rl.is_key_down(KEY_LEFT_SUPER),
            });
        }
    }
    out
}
pub fn measure_text(text: &str, font_size: i32) -> i32 {
    let c = std::ffi::CString::new(text).unwrap_or_default();
    unsafe { raylib::ffi::MeasureText(c.as_ptr(), font_size) }
}

pub fn draw_text_centered<D: RaylibDraw>(
    d: &mut D,
    text: &str,
    cx: i32,
    cy: i32,
    size: i32,
    color: Color,
) {
    let w = measure_text(text, size);
    d.draw_text(text, cx - w / 2, cy - size / 2, size, color);
}
